### PR TITLE
Embedded timestepper and driver for Richards

### LINKED ISSUE #68: Ability for SNES-based time integrator to run to a certain time and pause

### PR DESCRIPTION
**Pull request by ghammond86**
- Embeds the SNES and (not TS) time stepper within TDy
- Employs a driver to set up a simulation - had to use a wrapper to tdycore as otherwise there are intermingled dependencies
- The driving program initializes, runs to a point in time, destroys.  Run-to-time can be called many times.
- Fixes #68

### CODE REVIEW

**Comment by bishtgautam** [comment]
@ghammond86 How is one suppose to run `demo/richards/richards_driver`?

File: include/tdyts.h [hunk 1]
```diff
@@ -0,0 +1,25 @@
+#if !defined(TIMESTEPPER_H)
+#define TIMESTEPPER_H
+
+#include <petsc.h>
+
+typedef struct Timestepper *Timestepper;
+
+struct Timestepper {
+  SNES snes;
```
Comment (bishtgautam) [inline include/tdyts.h:9]: This structure would eventually also include `TS ts;`. So, we need an integer that identifies if `SNESSet/Solve` or `TSSet/Solve` would be called. So, how about adding an integer that identifies the type of timestepper (e.g. `PetscInt solver_type`).

File: src/tdydriver.c [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+  ierr = TDySetPorosityFunction(tdy,TDyPorosityFunctionDefault,PETSC_NULL);
+         CHKERRQ(ierr);
+  ierr = TDySetPermeabilityFunction(tdy,TDyPermeabilityFunctionDefault,PETSC_NULL);
+         CHKERRQ(ierr);
+  ierr = TDySetDiscretizationMethod(tdy,MPFA_O); CHKERRQ(ierr);
+  ierr = TDySetFromOptions(tdy); CHKERRQ(ierr);
+
+  ierr = TDyTimestepperCreate(&tdy->ts); CHKERRQ(ierr);
+  ierr = SNESCreate(PETSC_COMM_WORLD,&tdy->ts->snes); CHKERRQ(ierr);
+  ierr = TDySetSNESFunction(tdy->ts->snes,tdy); CHKERRQ(ierr);
```
Comment (bishtgautam) [inline src/tdydriver.c:22]: Adding a `switch (tdy->ts->solver_type)`

File: src/tdydriver.c [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+  ierr = SNESGetLineSearch(tdy->ts->snes,&linesearch); CHKERRQ(ierr);
+  ierr = SNESLineSearchSetPostCheck(linesearch,TDyRichardsSNESPostCheck,&tdy);
+         CHKERRQ(ierr);
+  ierr = SNESSetFromOptions(tdy->ts->snes); CHKERRQ(ierr);
+
+  switch (tdy->mode) {
+    case RICHARDS:
+      ierr = TDyRichardsInitialize(tdy); CHKERRQ(ierr);
+      break;
+    case TH:
```
Comment (bishtgautam) [inline src/tdydriver.c:34]: Add a default and report an error.

File: src/tdyts.c [hunk 4]
```diff
@@ -0,0 +1,89 @@
+#include <private/tdycoreimpl.h>
+#include <tdycore.h>
+#include <tdyts.h>
+#include <tdyio.h>
+
+PetscErrorCode TDyTimestepperCreate(Timestepper *_ts) {
+  Timestepper ts;
```
Comment (bishtgautam) [inline src/tdyts.c:7]: Let's only use `ts` for PETSc TS.

**Comment by ghammond86** [comment]
> @ghammond86 How is one suppose to run `demo/richards/richards_driver`?

I added run.sh that provides an example.

**Comment by bishtgautam** [comment]
@ghammond86 In the first comment of this PR, you can add 'Fixes #68'. So that issue will be automatically closed on this is PR is merged.

**Comment by ghammond86** [comment]
> I have made a few comments and have additional requests:
> 
> * update the `demo/makefile` so the `richards_driver.c` is compiled from the demo directory.

I do not follow this comment.  Do you mean update the regression_test/Makefile so that richard_driver.c is compiled from the test directroy.  TDYCORE_DIR/demo/richards/makefile is identical to the other demo makefiles except for the c source filename and targets.

> * add a new test for `richards_driver`.

**Comment by bishtgautam** [comment]
I meant updating the https://github.com/TDycores-Project/TDycore/blob/master/demo/makefile

File: src/tdyts.c (hunk 4 above)
Comment (ghammond86) [reply src/tdyts.c:7]: How about TimeIntegrator and 'ti'?

Comment (ghammond86) [reply src/tdyts.c:7]: BTW, I have already refactored to TDyTimestepper and timestepper.

Comment (bishtgautam) [reply src/tdyts.c:7]: `TDyTimeIntegrator` and `ti` sound good to me.

File: src/tdydriver.c [hunk 5]
```diff
@@ -17,22 +17,40 @@ PetscErrorCode TDyDriverInitializeTDy(TDy tdy) {
...
+      ierr = TDySetSNESFunction(tdy->ti->snes,tdy); CHKERRQ(ierr);
+      ierr = TDySetSNESJacobian(tdy->ti->snes,tdy); CHKERRQ(ierr);
+      SNESLineSearch linesearch;
+      ierr = SNESGetLineSearch(tdy->ti->snes,&linesearch); 
+             CHKERRQ(ierr);
+      ierr = SNESLineSearchSetPostCheck(linesearch,TDyRichardsSNESPostCheck,
+                                        &tdy); CHKERRQ(ierr);
+      ierr = SNESSetFromOptions(tdy->ti->snes); CHKERRQ(ierr);
+      break;
+    case TDyTS:
```
Comment (bishtgautam) [inline src/tdydriver.c:34]: Add an error that `TdyTS` is unsupported.

File: src/tdyti.c [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+          printf("Time step %d: time = %f dt = %f ni=%d li = %d\n",
+                 ti->istep,ti->time,ti->dt,nit,lit);
+        if (tdy->io->print_intermediate) {
+          ierr = TDyIOPrintVec(tdy->solution,"soln",ti->istep); 
+          CHKERRQ(ierr);
+        }
+        ierr = TDyTimeIntegratorUpdateDT(ti,sync_time); CHKERRQ(ierr);
+      }
+      break;
+    case TDyTS:
```
Comment (bishtgautam) [inline src/tdyti.c:85]: Add an error that `TdyTS` is unsupported.

Comment (ghammond86) [reply src/tdyti.c:85]: I have added error messages in tdydriver to catch the use of both TDyTS and TH.

**Pull request merged**